paper/ml.tex

```diff
@@ -2,7 +2,68 @@ \section{AI and Machine Learning Contributions}\label{ml-sec}
 
-On the other hand, we found that machine learning (ML) methods showed some promise of being able to heuristically predict the truth value of portions of implication graph, as we shall now discuss\footnote{For some discussion of other ML experiments performed during the project, see \url{https://leanprover.zulipchat.com/\#narrow/channel/458659-Equational/topic/Machine.20learning.2C.20first.20results}.}.
+On the other hand, we found that machine learning (ML) methods showed some promise of being able to heuristically predict the truth value of portions of the implication graph, as we shall now discuss\footnote{For some discussion of other ML experiments performed during the project, see \url{https://leanprover.zulipchat.com/\#narrow/channel/458659-Equational/topic/Machine.20learning.2C.20first.20results}.}.
+
+\subsection{Convolutional neural network model for the implication graph}
+
+To model the implication graph, we used a convolutional neural network (CNN). For each pair of equations $(p,q)$, the input of the CNN consisted on a character-level tokenization (no vectorization) of the two equations, and the output of the CNN was a yes/no label depending on whether $p$ implies $q$. The CNN processed the input data using a 5-layer architecture, each layer composed of a 1-D convolution, followed by batch normalization and rectified linear unit activation functions (\cite{Goodfellow-et-al-2016}). After the last convolutional layer, a flattening layer and a softmax activation function were used to obtain the output of the network, i.e., the prediction of the implication for the input pair of equations. Note that we trained our models with the 16/10/2024 version of the (infinite) implication graph, for which 362 of the hardest implications (less than 0,002\% of the total) were still unknown.
+
+\smallskip
+
+Prior to training the CNN, we divided the data into training (60\%),validation (20\%), and test (20\%) subsets. The CNN was implemented in TensorFlow 2.9.0 (\cite{tensorflow2015-whitepaper}) and trained on an NVIDIA 3080 Ti GPU with the following configuration: the binary cross-entropy as the loss function to minimize, the Adam method with an initial learning rate of $10^{-3}$ for the adjustment of network weights, a batch size of 1024 with random shuffling, a learning rate reduction by a factor of 2 after 15 epochs without improvement in the validation loss, and early stopping if no improvement occurred for 40 epochs, being the model with the lowest validation loss retained as the final CNN.
```

Comment:
This is a list of configs, but the last clause ("being the model.. retained") doesn't quite fit that pattern.